File: problems/78-subsets/solution.cpp
class Solution {
 private:
    void helper(vector<vector<int> > &res,
                vector<int> &out,
                vector<int> &nums,
                int start) {
        res.push_back(out);
        for (int i = start; i < nums.size(); ++i) {
            out.push_back(nums[i]);
            helper(res, out, nums, i + 1);
            out.pop_back();
        }
    }
 public:
    vector<vector<int> > subsets(vector<int> &nums) {
        vector<vector<int> > res;
        vector<int> out;
        helper(res, out, nums, 0);
        return res;
    }
};as

File: problems/78-subsets/solution.py
#
# @lc app=leetcode id=78 lang=python3
#
# [78] Subsets
#

# @lc code=start
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def dfs(nums, out, ret, start):
            ret.append(out)
            for i in range(start, len(nums)):
                dfs(nums, out + [nums[i]], ret, i + 1)
        if not nums:
            return []
        ret = []
        dfs(nums, [], ret, 0)
        return ret

'''
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.res = []
        self.helper(nums, 0, [])
        return self.res
      
    def helper(self, nums, start, out):
        self.res.append(out[:])
        for i in range(start, len(nums)):
            self.helper(nums, i + 1, out + [nums[i]])
'''
# @lc code=end

